src/target/target_flash.c

```diff
@@ -161,18 +161,18 @@ bool target_flash_write(target *t, target_addr_t dest, const void *src, size_t l
 	if (!target_enter_flash_mode(t))
 		return false;
 
+	target_flash_s *active_flash = target_flash_for_addr(t, dest);
 	bool ret = true; /* catch false returns with &= */
 	while (len) {
 		target_flash_s *f = target_flash_for_addr(t, dest);
 		if (!f)
 			return false;
 
 		/* terminate flash operations if we're not in the same target flash */
-		for (target_flash_s *target_f = t->flash; target_f; target_f = target_f->next) {
-			if (target_f != f) {
-				ret &= flash_buffered_flush(target_f);
-				ret &= flash_done(target_f);
-			}
+		if (f != active_flash) {
+			ret &= flash_buffered_flush(active_flash);
+			ret &= flash_done(active_flash);
+			active_flash = f;
```

Comment:
If we can figure out how to deal with incomplete blocks to handle the point below, then this would be fine.. the partial writes problem suggests this needs two levels of function - one to turn the partials into complete writes with vFlashDone calling `flash_buffered_flush()` + `flash_done()` to handle the final exit condition.. need to think on this carefully as you're right, we would do well to ensure this is always correct - a problem that #1231 actually also hits but we didn't realise fully at the time (early exiting is problematic).

Reply:
We didn't end up quite doing things how we talked about.. but what we have done should still provide the improvement needed. Please check that the logic now implemented by this PR doesn't introduce the hole mentioned in this review.